feat!: return metadata associated to private events

mverzilli opened the pull request:
Related to https://github.com/AztecProtocol/aztec-packages/issues/16245
Part of F-96

## New types

### Wallet interface
  - `PrivateEvent<T>`: bundles together an ABI decoded event and contextual metadata.
  - `PrivateEventFilter`: weaker version of `PrivateEvent` to represent filters. Takes some hints from Aztec Node's `LogFilter` for consistency.

### PXE
 - `PackedPrivateEvent`: private event contents in its Field array representation plus metadata.

## Quirks

Aztec Node's `LogFilter` establishes `fromBlock` to be inclusive, and `toBlock` exclusive, so I'm leaning towards respecting that convention here to keep a consistent approach. It results in some idiosyncratic choices when defining defaults. At a first approximation, I think defaulting to "returns events emitted in the latest known block" is reasonable, but open to other ideas.

nventuro commented on docs/docs/developers/docs/resources/migration_notes.md:
```diff
@@ -437,6 +325,66 @@ Return value of `getNotes` used to be defined as `Promise<UniqueNote[]>` and now
 `NoteDao` is mostly a super-set of `UniqueNote` but it doesn't contain a `recipient`.
 Having the recipient in the return value has been redundant as the same outcome can be achieved by populating the `scopes` array in `NoteFilter` with the `recipient` value.
 
+#### Changes to `getPrivateEvents`
```
This is very good

benesjan commented on yarn-project/aztec.js/src/wallet/wallet.ts:
```suggestion
  /** Transaction in which the events were emitted. */
```

benesjan commented on yarn-project/pxe/src/storage/private_event_data_provider/private_event_data_provider.ts:
```suggestion
    filter: PrivateEventStoreFilter,
```

Having it be plural given that the type is called `PrivateEventStoreFilter` seems weird.

benesjan commented on yarn-project/pxe/src/storage/private_event_data_provider/private_event_data_provider.ts:
I had to look up the difference between `PrivateEventFilter` and `PrivateEventStoreFilter` and I think the name could be better here.

Given that in PXE there is `validateGetPrivateEventsFilters` function maybe we could call it `ValidatedPrivateEventsFilter`?

Might also make sense to make this a class and move the functionality of `validateGetPrivateEventsFilters` to `fromPrivateEventFilter` function to move some complexity out of the `pxe.ts` file which is already a bit too large (If you decide to do this would put it in a separate file).

Also would put here a small comment that it's like `PrivateEventFilter` but with populated and validated blocks.

mverzilli replied:
Good idea. Regarding the name, the more appropriate one would be PrivateEventDataProviderFilter but it seemed too much 😁. Will see if I'm more inspired tomorrow.

nventuro replied:
>PrivateEventDataProviderFilter

This seems fine to me. If you're reading code this deep in the database then being very explicit about what's a database filter and what's a user input filter sounds like a great idea.

benesjan commented on yarn-project/pxe/src/storage/private_event_data_provider/private_event_data_provider.ts:
```suggestion
   * the transaction and block the event was included in.
```

benesjan commented on yarn-project/pxe/src/pxe.test.ts:
If the `validateGetPrivateEventsFilters` gets moved on to the filter type I think we will be able to afford to move this only to unit tests of the type.

mverzilli replied:
I ended up introducing an ad-hoc type to transform and validate from PXE's filter type to the data provider filter type, and moving all related unit tests there. Eventually we might decompose that further so we reuse block range validation and defaults logic throughout the codebase, but that will be future work

benesjan commented on yarn-project/pxe/src/pxe.ts:
```suggestion
  async #validateGetPrivateEventsFilter(filter: PrivateEventFilter): Promise<PrivateEventStoreFilter> {
```

The argument is singular so would keep the name consistent.

benesjan commented on yarn-project/pxe/src/pxe.ts:
I see that `eth_getLogs` behaves differently:

<img width="910" height="176" alt="Image" src="https://github.com/user-attachments/assets/f4c281f3-5a7c-406f-a5d8-81c6e2cceffe" />

I think we should change the behavior in both node and here such that Ethereum's convention is followed. Feels like this is the last good time when we can afford to do this.

WDYT?

(Obviously that would be a change for another PR)

mverzilli replied:
I'm fine either way as long as we adopt one everywhere. I do have a slight preference towards the Ethereum style.

I'm still working on related event/log querying features so I can tackle that in a different PR (after checking with Palla that we aren't causing much trouble)

mverzilli replied:
@spalladino summoning you here

mverzilli replied:
Discussed with Palla an he's fine with this, I'm deferring this another PR though, given it will ripple through the codebase (tracking through F-198)

benesjan commented on yarn-project/stdlib/src/block/in_block.ts:
Is this a forgotten TODO or do you plan to tackle it in a followup PR?

mverzilli replied:
Forgotten, good catch

mverzilli replied:
Just realized this problem is legacy, so removed the TODO for now, but tracking in linear (F-199)

benesjan reviewed (dismissed):
Overall very nice and will be happy to approve once we discuss my feedback.

Great job!

benesjan commented:
Accidentally approved instead of just commenting. Please just ping me if you will want my feedback again

mverzilli commented:
> Accidentally approved instead of just commenting. Please just ping me if you will want my feedback again

No worries it's now blocked (presumably by the branded block number pr) so your feedback won't go anywhere until I can solve the conflicts tomorrow 🥲.

Thanks for reviewing btw!

nventuro commented on yarn-project/aztec.js/src/wallet/wallet.ts:
```suggestion
  /** The address of the contract that emitted the events. */
  contractAddress: AztecAddress;
```

'filter by' is implied because this is a filter. I'd focus instead of specifying what each param means.

nventuro commented on yarn-project/aztec.js/src/wallet/wallet.ts:
```suggestion
 * Filter options when querying private events.
```

nventuro commented on yarn-project/aztec.js/src/wallet/wallet.ts:
Events don't have recipients, they are just emitted. The _messages_ containing the events are delivered to some recipient. When we process a message, we know the recipient of the message. If processing a message results in an event being found, we tell PXE of the recipient that got the message from where that event originates. PXE then chooses to _scope_ the event to that recipient, meaning queries that do not include that account in its scopes do not return the event - mimicking other accounts not 'knowing' about the event.

So this parameter is really the accounts that are in scope when querying events. Recipient and scope are still a bit intertwined and in some cases confused for one another (since they are quite similar), but we should try to improve this.

This also begs the question - what happens if the same event is in multiple scopes? We should have a deduplication step in PXE.

mverzilli replied:
> This also begs the question - what happens if the same event is in multiple scopes? We should have a deduplication step in PXE.

Mmm, interesting. With the current changes you would see duplicates of the event contents, but differences in metadata: `recipient` (soon to be `scope`) would change. Alternatively, we could remove that field from metadata and indeed deduplicate. But if we go that route it doesn't make sense to accept an array here and the query should always be limited to one address.

mverzilli replied:
Modified language and names to refer to scopes instead of recipients. Re: deduplication, when looking into it I found that there's no risk of duplicates basically because the data provider as a whole is not doing what we need it to do, and we mapped what needs to be done there on a new linear project so leaving it at that for now.

nventuro commented on yarn-project/aztec.js/src/wallet/wallet.ts:
This seems like a bad default? In most cases this would only return events from the very last block.

benesjan replied:
I was also surprised by this but Ethereum clients also seem to default only to the latest block.

mverzilli replied:
The rationale for this is to default to a limited range, I didn't want to encourage `[0, infinity)` filters. 

It is kind of moot at the moment because `PrivateEventDataProvider#getPrivateEvents` does a full scan anyway, but since this is an externally facing API and eventually we will want to honor backwards compatibility, it might still make sense to be conservative by default.

That said, the node defaults to the first block, so maybe we should stick to that:

```typescript
const start =
      filter.afterLog?.blockNumber ?? Math.max(filter.fromBlock ?? INITIAL_L2_BLOCK_NUM, INITIAL_L2_BLOCK_NUM);
```

That also touches on a discussion we have with @benesjan in somewhere else in this PR: do we want to mimic what Ethereum does here? If so we would need to change the behavior of the node. If otherwise we stick to how the node works right now, I can indeed apply this same logic.

nventuro replied:
I think it's more important to mirror the behavior of our node. We can always change what our node does, but the wallet should not diverge from it.

nventuro commented on yarn-project/aztec.js/src/wallet/wallet.ts:
I'd expect queries from block 0 to `toBlock` in this case tbh.

benesjan replied:
INITIAL_L2_BLOCK_NUM is set to 1 so I would default to that constant.

It's not zero because it made implementation of Rollup contracts easier. I know longer remember why.

nventuro commented on yarn-project/aztec.js/src/wallet/wallet.ts:
I don't think we should return the recipient of the event to the app. It looks like an unnecessary privacy leak. I'd just return the event data and tx. See my message above re. scopes. The fact that the recipient is here hints at us not doing any deduplication (because we'd consider the same event processed by different recipients to be different).

nventuro commented on yarn-project/aztec.js/src/wallet/wallet.ts:
Given the existence of `EventMetadataDefinition`, `PrivateEventMetadata` sounds like it might lead to confusion. But given my comment above re. recipient, we likely can just remove this type.

nventuro commented on yarn-project/pxe/src/pxe.ts:
This is evidence of the need for PXE to handle contract syncing as a first class concept.

mverzilli replied:
Agree, ideally this should happen only at a single point in PXE's "outer loop" and not at random points spread through the codebase

nventuro commented on yarn-project/pxe/src/pxe.ts:
Agree with Jan on the intuition of moving this into the events file and out of pxe, though I'm a bit torn. On the one hand, this is about creating the object needed by the events, and it is that one that understands what is required re. validation. On the other, the validation and selection of defaults is a key concern of PXE as it informs its user interface. So I don't really have a suggestion re. this, just pointing this out.

It does seem odd however that this file would be doing both this and orchestrating complex tasks such as proving (eg. `PXE.proveTx`). Perhaps what's wrong is to have such complex behavior here, and this class should be more of an input validation + transformation layer that then delegates to more low level components (the data providers, some proving orchestrator, etc.).

mverzilli replied:
I think there's two layers to decompose this to:

1. Block range filtering logic: you'll notice most of the discussion in this PR revolves around what defaults we choose and the intention of making those defaults coherent throughout the stack. Given that, it seems sensible to codify it in a single place and try to use it everywhere. I would even suggest putting it in stdlib.
2. PXE input sanitization: this makes sense to extract to write lighter unit tests as Jan suggested, and it could use the module I describe in point 1. However, it would still be ad-hoc for PXE use and not aim at reuse.

mverzilli replied:
Follow up: 

https://github.com/AztecProtocol/aztec-packages/pull/18234#discussion_r2592234967

and 

https://github.com/AztecProtocol/aztec-packages/pull/18234/files#r2592224639

nventuro reviewed (commented):
Left some comments re. some interface choices and architectural bits. The low level details such as tests etc. look good, so I only focused on these things.

nventuro commented on docs/docs/developers/docs/resources/migration_notes.md:
```diff
@@ -355,20 +355,20 @@ export type InTx = InBlock & {
-};
-
 export type PrivateEvent<T> = {
   event: T;
-  metadata: PrivateEventMetadata;
+  metadata: InTx;
 };
```
Would it make sense to do 

```ts
export type PrivateEvent<T> = {
  event: T;
} & InTx;
```

Like you did for `InBlock`? Just curious re. the approach here.

mverzilli replied:
`T` is eventually an ABI decoded type defined in user land, so I didn't want to expose us to shadowing/namespacing issues

mverzilli commented on yarn-project/pxe/src/events/index.ts:
Note: eventually I would like to organize the PXE code more around vertical features (events, capsules, etc) instead of layers (stores, validators, etc). So here we would also have PrivateDataEventProvider and related files for example. 

I'm not doing it now because it would blast this PR with changes unrelated to its intent

benesjan commented on yarn-project/pxe/src/events/private_event_filter_validator.ts:
```diff
@@ -0,0 +1,45 @@
+    // the first block to the latest known block.
+    if (!fromBlock || !toBlock) {
+      const lastKnownBlock = await this.syncDataProvider.getBlockNumber();
+      fromBlock = fromBlock ?? 0;
```
We have `INITIAL_L2_BLOCK_NUM` constant defined so think it would make sense to use here.

benesjan commented on yarn-project/aztec.js/src/wallet/wallet.ts:
```diff
@@ -129,19 +132,48 @@ export type BatchResults<T extends readonly BatchedMethod<keyof BatchableMethods
+  txHash?: TxHash;
+  /** The block number from which to start fetching events (inclusive).
+   * Optional. If provided, it must be greater or equal than 0.
+   * Defaults to 0.
```
Would use here the `INITIAL_L2_BLOCK_NUM` but would also accept 0 on the input as valid to not be unnecessarily annoying for devs and It doesn't really matter.

benesjan commented on yarn-project/pxe/src/events/private_event_filter_validator.test.ts:
```diff
@@ -0,0 +1,122 @@
+    expect(dataProviderFilter).toEqual({
+      contractAddress,
+      scopes: [scope],
+      fromBlock: BlockNumber(1),
```
```suggestion
      fromBlock: INITIAL_L2_BLOCK_NUM
```

benesjan commented on yarn-project/pxe/src/events/private_event_filter_validator.test.ts:
```diff
@@ -0,0 +1,122 @@
+    });
+  });
+
+  it('toBlock without fromBlock defaults to [0, toBlock)', async () => {
```
```suggestion
  it('toBlock without fromBlock defaults to [INITIAL_L2_BLOCK_NUM, toBlock)', async () => {
```

benesjan commented on yarn-project/pxe/src/events/private_event_filter_validator.test.ts:
```diff
@@ -0,0 +1,122 @@
+    let dataProviderFilter = await validator.validate({
+      contractAddress,
+      scopes: [scope],
+      fromBlock: BlockNumber(1),
```
```suggestion
      fromBlock: INITIAL_L2_BLOCK_NUM,
```

Would just use here the constant here and below given that we have it available.

benesjan commented on yarn-project/pxe/src/events/private_event_filter_validator.ts:
```diff
@@ -0,0 +1,46 @@
+    // the first block to the latest known block.
+    if (!fromBlock || !toBlock) {
+      const lastKnownBlock = await this.syncDataProvider.getBlockNumber();
+      fromBlock = fromBlock ?? BlockNumber(1);
```
```suggestion
      fromBlock = fromBlock ?? INITIAL_L2_BLOCK_NUM;
```

benesjan commented on yarn-project/pxe/src/events/private_event_filter_validator.ts:
```diff
@@ -0,0 +1,46 @@
+      throw new Error('At least one scope is required to get private events');
+    }
+
+    if (fromBlock < 1) {
```
```suggestion
    if (fromBlock < INITIAL_L2_BLOCK_NUM) {
```

mverzilli replied:
not applying this one, because of this: https://github.com/AztecProtocol/aztec-packages/pull/18234#discussion_r2599741080

benesjan commented on yarn-project/pxe/src/events/private_event_filter_validator.ts:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    if (fromBlock < 1) {
+      throw new Error('fromBlock must be greater than or equal to 1');
```
```suggestion
      throw new Error(`fromBlock must be greater than or equal to ${INITIAL_L2_BLOCK_NUM}`);
```

mverzilli replied:
I think it's best to be consistent here, if we accept lower values then we shouldn't contradict that in the error message

benesjan commented on yarn-project/pxe/src/events/private_event_filter_validator.ts:
```diff
@@ -0,0 +1,46 @@
+      throw new Error('fromBlock must be greater than or equal to 1');
+    }
+
+    if (toBlock < 1) {
```
```suggestion
    if (toBlock < INITIAL_L2_BLOCK_NUM) {
```

benesjan commented on yarn-project/pxe/src/events/private_event_filter_validator.ts:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    if (toBlock < 1) {
+      throw new Error('toBlock must be greater than or equal to 1');
```
```suggestion
      throw new Error(`toBlock must be greater than or equal to ${INITIAL_L2_BLOCK_NUM}`);
```

benesjan commented on yarn-project/pxe/src/storage/private_event_data_provider/private_event_data_provider.test.ts:
```diff
@@ -16,211 +17,175 @@ const getRandomMsgContent = () => {
-      blockNumber,
-      blockHash,
-    );
-    await privateEventDataProvider.storePrivateEventLog(
-      contractAddress,
-      recipient,
-      eventSelector,
-      msgContent,
-      txHash,
-      eventCommitmentIndex,
-      blockNumber,
-      blockHash,
-    );
```
Given that we now don't store the event twice this test case seems to be broken.

mverzilli replied:
for now I'll leave this as is, PrivateEventDataProvider needs to be rewritten from scratch, but the current test suite at least documents what it was trying to implement. eg: if I hadn't seen this test here I wouldn't have raised the concern that the implementation was looking suspicious

benesjan reviewed (approved):
LGTM - just that one test needs fixing and think it makes sense to use the INITIAL_L2_BLOCK_NUM in places. Thanks

mverzilli committed:
feat!: return metadata associated to private events

Related to https://github.com/AztecProtocol/aztec-packages/issues/16245
Part of F-96

## New types

### Wallet interface
  - `PrivateEvent<T>`: bundles together an ABI decoded event and contextual metadata.
  - `PrivateEventFilter`: weaker version of `PrivateEvent` to represent filters. Takes some hints from Aztec Node's `LogFilter` for consistency.

### PXE
 - `PackedPrivateEvent`: private event contents in its Field array representation plus metadata.

## Quirks

Aztec Node's `LogFilter` establishes `fromBlock` to be inclusive, and `toBlock` exclusive, so I'm leaning towards respecting that convention here to keep a consistent approach. It results in some idiosyncratic choices when defining defaults. At a first approximation, I think defaulting to "returns events emitted in the latest known block" is reasonable, but open to other ideas.
